pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go

```diff
@@ -0,0 +1,380 @@
+}
+
+// Admit rejects all pods if node is shutting
+func (m *managerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	nodeShuttingDown := m.ShutdownStatus() != nil
+
+	if nodeShuttingDown {
+		return lifecycle.PodAdmitResult{
+			Admit:   false,
+			Reason:  nodeShutdownNotAdmittedReason,
+			Message: nodeShutdownNotAdmittedMessage,
+		}
+	}
+	return lifecycle.PodAdmitResult{Admit: true}
+}
+
+// setMetrics sets the metrics for the node shutdown manager.
+func (m *managerImpl) setMetrics() {
+	if m.enableMetrics && m.storage != nil {
+		sta := state{}
+		err := m.storage.Load(&sta)
+		if err != nil {
+			m.logger.Error(err, "Failed to load graceful shutdown state")
+		} else {
+			if !sta.StartTime.IsZero() {
+				metrics.GracefulShutdownStartTime.Set(timestamp(sta.StartTime))
+			}
+			if !sta.EndTime.IsZero() {
+				metrics.GracefulShutdownEndTime.Set(timestamp(sta.EndTime))
+			}
+		}
+	}
+}
+
+// Start starts the node shutdown manager and will start watching the node for shutdown events.
+func (m *managerImpl) Start() error {
+	m.logger.V(1).Info("Shutdown manager get started")
+
+	_, err := m.start()
+
+	if err != nil {
+		return err
+	}
+
+	service.SetPreShutdownHandler(m)
+
+	m.setMetrics()
+
+	return nil
+}
+
+// Check the return value, maybe return error only
+func (m *managerImpl) start() (chan struct{}, error) {
+	// Process the shutdown only when it is running as a windows service
+	isServiceInitialized := service.IsServiceInitialized()
+	if !isServiceInitialized {
+		return nil, errors.Errorf("%s is NOT running as a Windows service", serviceKubelet)
+	}
+
+	// Update the registry key to add the kubelet dependencies to the existing order
+	mgr, err := mgr.Connect()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not connect to service manager")
+	}
+	defer mgr.Disconnect()
+
+	s, err := mgr.OpenService(serviceKubelet)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not access service %s", serviceKubelet)
+	}
+	defer s.Close()
+
+	preshutdownInfo, err := service.QueryPreShutdownInfo(s.Handle)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not query preshutdown info")
+	}
+
+	config, err := s.Config()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not access config of service %s", serviceKubelet)
+	}
+
+	// Open the registry key
+	key, err := registry.OpenKey(registry.LOCAL_MACHINE, shutdownOrderRegPath, registry.QUERY_VALUE|registry.SET_VALUE)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not access registry")
+	}
+	defer key.Close()
+
+	// Read the existing values
+	existingOrders, _, err := key.GetStringsValue(shutdownOrderStringValue)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not access registry value %s", shutdownOrderStringValue)
+	}
+
+	// Add the kubelet dependencies to the existing order
+	newOrders := addToExistingOrder(config.Dependencies, existingOrders)
+	err = key.SetStringsValue("PreshutdownOrder", newOrders)
```

Comment:
do we have to set it if it's the same as before?

Reply:
yes, this is the current logic, otherwise we will need to create new slices and do the comparison, I'm thinking maybe a blind copy is easier and performant